azurelinuxagent/ga/extension_telemetry.py

```diff
@@ -0,0 +1,385 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import json
+import os
+import re
+import threading
+from collections import defaultdict
+
+import azurelinuxagent.common.logger as logger
+from azurelinuxagent.common import conf
+from azurelinuxagent.common.event import EVENTS_DIRECTORY, TELEMETRY_LOG_EVENT_ID, \
+    TELEMETRY_LOG_PROVIDER_ID, add_common_params_to_extension_event, add_event, WALAEventOperation, add_log_event
+from azurelinuxagent.common.exception import InvalidExtensionEventError
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.telemetryevent import TelemetryEventList, TelemetryEvent, TelemetryEventParam, \
+    GuestAgentGenericLogsSchema
+from azurelinuxagent.ga.exthandlers import HANDLER_NAME_PATTERN
+from azurelinuxagent.ga.periodic_operation import PeriodicOperation
+
+
+def get_extension_telemetry_handler(protocol_util):
+    # Reuse the same protocol_util as the ExtHandler class
+    return ExtensionTelemetryHandler(protocol_util)
+
+class ExtensionEventSchema(object):
+    """
+    Class for defining the schema for Extension Events.
+    """
+    Version = "Version"
+    Timestamp = "Timestamp"
+    TaskName = "TaskName"
+    EventLevel = "EventLevel"
+    Message = "Message"
+    EventPid = "EventPid"
+    EventTid = "EventTid"
+    OperationId = "OperationId"
+
+class ExtensionTelemetryHandler(object):
+    """
+    This Handler takes care of fetching the Extension Telemetry events from the {extension_events_dir} and sends it to
+    Kusto for advanced debuggability.
+    """
+
+    EXTENSION_EVENT_COLLECTION_PERIOD = datetime.timedelta(minutes=5)
```

Comment:
Except for the limits, made everything private. Since we need to mock the limits for unit tests, I didnt make them private. In general should we consider making such variables private if we need to mock them? Is there a general guideline around that?

Reply:
Had an offline sync, will make all variables private.